lib/Cake/Utility/Hash.php

```diff
@@ -756,18 +756,15 @@ public static function dimensions(array $data) {
  * number of dimensions in a mixed array.
  *
  * @param array $data Array to count dimensions on
+ * @param int $count current depth count for this interation of the function
  * @return int The maximum number of dimensions in $data
  * @link http://book.cakephp.org/2.0/en/core-utility-libraries/hash.html#Hash::maxDimensions
  */
-	public static function maxDimensions(array $data) {
-		$depth = array();
+	public static function maxDimensions($data, $count = 0) {
+		$depth = array($count);
 		if (is_array($data) && reset($data) !== false) {
 			foreach ($data as $value) {
-				if (is_array($value)) {
-					$depth[] = self::dimensions($value) + 1;
-				} else {
-					$depth[] = 1;
-				}
+				$depth[] = static::maxDimensions($value, $count + 1 );
```

Comment:
I am not sure if this is the most efficient way.  I pieced the code together from the deprecated Set::countDim function primarily because the current Hash::maxDimensions function is not returning the correct value.  I discovered this while trying to migrate from the deprecated Set functions to the new Hash functions.